services/src/main/java/com/linkedin/kmf/services/TopicRebalancer.java

```diff
@@ -0,0 +1,427 @@
+    for (PartitionInfo javaPartitionInfo : partitionInfoList) {
+      scalaPartitionInfoSet.add(new TopicAndPartition(_topic, javaPartitionInfo.partition()));
+    }
+    PreferredReplicaLeaderElectionCommand.writePreferredReplicaElectionData(zkUtils, scalaPartitionInfoSet);
+  }
+
+  protected void addPartitions(ZkUtils zkUtils, int addPartitionCount) {
+    AdminUtils.addPartitions(zkUtils, _topic, addPartitionCount, null, false);
+  }
+
+  protected ZkUtils createZooKeeperUtils() {
+    if (_zkUtils == null) {
+      _zkUtils = ZkUtils.apply(_zkConnect, ZK_SESSION_TIMEOUT_MS + _scheduleDurationMs, ZK_CONNECTION_TIMEOUT_MS + _scheduleDurationMs,
+          JaasUtils.isZkSecurityEnabled());
+    }
+    return _zkUtils;
+  }
+
+
+  /**
+   * @return a value other than RebalanceCondition.OK if the monitored topic is no longer distributed evenly with respect
+   * to the given parameters.   Will not return null.
+   */
+  private RebalanceCondition monitoredTopicNeedsRebalance() {
+
+    ZkUtils zkUtils = createZooKeeperUtils();
+    List<PartitionInfo> partitionInfoList = partitionInfoForMonitoredTopic();
+    Collection<Broker> brokers = scala.collection.JavaConversions.asJavaCollection(zkUtils.getAllBrokersInCluster());
+    return monitoredTopicNeedsRebalance(partitionInfoList, brokers, _rebalanceThreshold);
+  }
+
+
+  private List<PartitionInfo> partitionInfoForMonitoredTopic() {
+    ZkUtils zkUtils = createZooKeeperUtils();
+    scala.collection.mutable.ArrayBuffer<String> topicList = new scala.collection.mutable.ArrayBuffer<>();
+    topicList.$plus$eq(_topic);
+    scala.collection.Map<Object, scala.collection.Seq<Object>> partitionAssignments =
+        zkUtils.getPartitionAssignmentForTopics(topicList).apply(_topic);
+    List<PartitionInfo> partitionInfoList = new ArrayList<>();
+    scala.collection.Iterator<scala.Tuple2<Object, scala.collection.Seq<Object>>> it = partitionAssignments.iterator();
+    while (it.hasNext()) {
+      scala.Tuple2<Object, scala.collection.Seq<Object>> scalaTuple = it.next();
+      Integer partition = (Integer) scalaTuple._1();
+      Node leader = new Node((Integer) zkUtils.getLeaderForPartition(_topic, partition).get(), "", -1);
+      Node[] replicas = new Node[scalaTuple._2().size()];
+      for (int i = 0; i < replicas.length; i++) {
+        Integer brokerId = (Integer) scalaTuple._2().apply(i);
+        replicas[i] = new Node(brokerId, "", -1);
+      }
+      partitionInfoList.add(new PartitionInfo(_topic, partition, leader, replicas, null));
+    }
+
+    return partitionInfoList;
+  }
+
+
+  /**
+   * This returns RebalanceCondition.OK unless one of the following conditions are met in this order:
+   * <ul>
+   *   <li> The minimum number of total monitored partitions falls below floor(brokers * partitionThreshold).
+   *        Returns PartitionsLow</li>
+   *   <li> One or more brokers does not have a monitored partition or falls below the minimum number of monitored
```

Comment:
I am not sure we need to check that every broker has at least a minimum number of partitions. We can discuss offline.